parquet-hadoop/src/main/java/org/apache/parquet/crypto/FileDecryptionProperties.java

```diff
@@ -0,0 +1,300 @@
+  private final AADPrefixVerifier aadPrefixVerifier;
+  private final Map<ColumnPath, ColumnDecryptionProperties> columnPropertyMap;
+  private final boolean checkPlaintextFooterIntegrity;
+  private final boolean allowPlaintextFiles;
+
+  private boolean utilized;
+
+  private FileDecryptionProperties(byte[] footerKey, DecryptionKeyRetriever keyRetriever,
+      boolean checkPlaintextFooterIntegrity,  byte[] aadPrefix, AADPrefixVerifier aadPrefixVerifier,
+      Map<ColumnPath, ColumnDecryptionProperties> columnPropertyMap, boolean allowPlaintextFiles) {
+
+    if ((null == footerKey) && (null == keyRetriever) && (null == columnPropertyMap)) {
+      throw new IllegalArgumentException("No decryption properties are specified");
+    }
+    if ((null != footerKey) && 
+        !(footerKey.length == 16 || footerKey.length == 24 || footerKey.length == 32)) {
+      throw new IllegalArgumentException("Wrong footer key length " + footerKey.length);
+    }
+    if ((null == footerKey) && checkPlaintextFooterIntegrity && (null == keyRetriever)) {
+      throw new IllegalArgumentException("Can't check footer integrity with null footer key and null key retriever");
+    }
+
+    this.footerKey = footerKey;
+    this.checkPlaintextFooterIntegrity = checkPlaintextFooterIntegrity;
+    this.keyRetriever = keyRetriever;
+    this.aadPrefix = aadPrefix;
+    this.columnPropertyMap = columnPropertyMap;
+    this.aadPrefixVerifier = aadPrefixVerifier;
+    this.allowPlaintextFiles = allowPlaintextFiles;
+    this.utilized = false;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private byte[] footerKeyBytes;
+    private DecryptionKeyRetriever keyRetriever;
+    private byte[] aadPrefixBytes;
+    private AADPrefixVerifier aadPrefixVerifier;
+    private Map<ColumnPath, ColumnDecryptionProperties> columnPropertyMap;
+    private boolean checkPlaintextFooterIntegrity;
+    private boolean plaintextFilesAllowed;
+
+    private Builder() {
+      this.checkPlaintextFooterIntegrity = CHECK_SIGNATURE;
+      this.plaintextFilesAllowed = ALLOW_PLAINTEXT_FILES;
+    }
+
+    /**
+     * Set an explicit footer key. If applied on a file that contains footer key metadata - 
+     * the metadata will be ignored, the footer will be decrypted/verified with this key.
+     * If explicit key is not set, footer key will be fetched from key retriever.
+     * 
+     * With explicit keys or AAD prefix, new encryption properties object must be created for each encrypted file.
+     * Explicit encryption keys (footer and column) are cloned.
+     * Upon completion of file reading, the cloned encryption keys in the properties will 
+     * be wiped out (array values set to 0).
+     * Caller is responsible for wiping out the input key array. 
+     * @param footerKey Key length must be either 16, 24 or 32 bytes. 
+     */
+    public Builder withFooterKey(byte[] footerKey) {
+      if (null == footerKey) {

```

Comment:
here, its not a validation.       _if (null == footerKey) { return this; }_
 a convenience that allows a user to set null instead of using checks when constructing the builder and deciding which "with**" methods to call and which to skip. Particularly useful for columns.